### ISSUE TITLE
PDF export: 'include::' -> Unresolved directive in <stdin> (path trouble)

### ISSUE THREAD

**Issue by SjoerdV** [asciidoctor/asciidoctor-vscode#700]
I tried to include a file using the relative path (seen from the workspace dir). 
```adoc
include::MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc[leveloffset=1]
```
Works fine in the VSCode preview. 

But delivers this message inside the exported PDF:
```error
Unresolved directive in <stdin> - include::MyCorp_-
_Office_Automation_Notebook/Support/Procedures_-
_50_shared_glossary.adoc[leveloffset=1]
```
Note there are `-_` constructions in the file/path name and these seem to be parsed incorrectly, in the error a newline is generated... I can also not escape the path with `++<filepath>++` this will break both the preview and the pdf. I also tried to make the path an appendix variable like so:
```adoc
:appendix-a: MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc
:appendix-b: ++MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc++

include::{appendix-a}[leveloffset=1]
include::{appendix-b}[leveloffset=1]
```
Only `appendix-a` showed correctly in the Preview. Neither worked in the pdf export (same error as above)

Maybe there is some other path escaping possible here?

VSCode (Linux): 1.74.3
AsciiDoc: v2.9.8

**Comment by ggrossetie** [comment]
Could you please share a screenshot or the PDF produced? I believe you are using `asciidoctor-pdf` command-line to generate the PDF? You might need to set the `base_dir` https://docs.asciidoctor.org/asciidoctor/latest/cli/man1/asciidoctor/#security-settings

**Comment by SjoerdV** [comment]
Asciidoctor PDF 2.3.4 using Asciidoctor 2.0.18 [https://asciidoctor.org]

@ggrossetie thanks for your suggestion, I added the `base_dir` in vscode settings, but same error appears. Also tried it as an inline setting in the document. this is the resulting PDF (in the error message it is missing an underscore just after the word 'Procedures' in the error message (or maybe just the error is being parsed instead of the real path, I don't know)

![2023-02-05-001](https://user-images.githubusercontent.com/13275320/216843234-b01bcbb1-1d1a-4670-b769-d9b916e3de1c.png)

these are the pdf related vscode settings:

![2023-02-05-002](https://user-images.githubusercontent.com/13275320/216843418-553fe230-84d1-4dc8-a3ac-6dcdebbcc749.png)

**Comment by SjoerdV** [comment]
I just found out that when I change the file reference to include parent path statements back to the workspace root folder:
```
include::../../MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc[leveloffset=1]
```
The PDF is now rendered correctly, except now the vscode preview does not show the include.

Option: I am perfectly fine adding the parent path references like above, so maybe it can be supported by the js preview?
Alternative Option: adding `:base_dir: ../../` to the document does nothing at the moment, but if it could be used for the pdf export it would also be fine.

What would you suggest @ggrossetie ?

**Comment by SjoerdV** [comment]
if I use a snippet as such I can have best of both worlds (preview will render correctly AND pdf export works)... but it might be a little too much overhead ;-)
```asciidoc
:external-snippet: MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc
ifdef::env-vscode[]
include::{external-snippet}[leveloffset=1] 
endif::[]
ifdef::backend-pdf[]
include::../../{external-snippet}[leveloffset=1]
endif::[]
```

**Comment by ggrossetie** [comment]
So it's a base directory issue, could you please try with the preversion 3.0.3? I made some changes to how `asciidoctor-pdf` is called.

**Comment by ggrossetie** [comment]
Can you please share your directory structure? Also `base_dir` can only be overriden from the CLI (or API) so you will need to configure `--base-dir=/path/to/my/basedir `.

> If the include directive is used in the primary (top-level) document, relative paths are resolved relative to the base directory. (The base directory defaults to the directory of the primary document and can be overridden from the CLI or API).

Does `MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc` is in the same directory as your primary document?

> Just a thought: wouldn't it be a lot easier if the --base-dir argument for asciidoctor-pdf could be (automatically) set same as the VSCode workspace directory.

It depends. Some users will prefer to resolve relative paths from the directory of the primary document and others from the VSCode workspace directory.
We might need to add an option in the settings.

**Comment by SjoerdV** [comment]
So this is my tree.
```
├── MyCorp_-_Office_Automation_Notebook
│   ├── [More Dirs]
│   ├── Features
│   │   ├── Features_-_50_shared_summary.adoc
│   └── Support
│   │   ├── Procedures_-_01_Introduction.adoc
│   │   └── Procedures_-_50_shared_glossary.adoc
├── resources
│   ├── css
│   ├── fonts
│   ├── media
│   │   └── icons
│   │── themes
│   └── default_settings.adoc
└── vscode_notebook
    └── pyaes
```

**all documents** in the 'MyCorp_-_Office_Automation_Notebook' structure  will **always** include the `resources/default_settings.adoc`. These contain all the generic settings.

Additionally, **any document** in the `MyCorp_-_Office_Automation_Notebook` structure **may** include appendices/addenda (usually from the same folder the main document is in., but these may come from other folders as well)

In this case the document `Procedures_-_01_Introduction.adoc` includes `default_settings.adoc`,`Procedures_-_50_shared_glossary.adoc` and `Features_-_50_shared_summary.adoc`

**Comment by ggrossetie** [comment]
I think we made a mistake in the VS code extension since includes should be resolved relative to the current document.

> In this case the document Procedures_-_01_Introduction.adoc includes default_settings.adoc,Procedures_-_50_shared_glossary.adoc and Features_-_50_shared_summary.adoc

The correct syntax is:

**Procedures_-_01_Introduction.adoc**
```
include::../resources/default_settings.adoc[]

include::Procedures_-_50_shared_glossary.adoc[]

include::../Features/Features_-_50_shared_summary.adoc[]
```

Otherwise, we will get a different result when using the CLI. I asked @mojavelinux about this issue on https://asciidoctor.zulipchat.com/#narrow/stream/279642-users/topic/Base.20directory.20and.20IDE (please feel free to add anything I may have missed in this thread)

**Comment by SjoerdV** [comment]
Beforehand: In your examples there should be one more level of 'parent' to match my example tree, so:
```
include::../../resources/default_settings.adoc[]

include::Procedures_-_50_shared_glossary.adoc[]

include::../../Features/Features_-_50_shared_summary.adoc[]
```
Things I learned and recommend so far:
1. I have reverted back to the release version, because the preview had too many issues (with toc and preview styles not loading period, whatever I did...) -> what follows is based on the current release 2.9.8 version
1. Taking a step back and looking at all the pathing issues (themes, styles, images, includes,etc.), what seems to be making this very confusing is the actual setting that was created to alleviate 'pathing' issues -> `asciidoc.useWorkspaceRoot` (in the preview it is renamed, but still present). This setting seems to only works for the 'JS' parts of the entire workflow and not for 'ascidoctor-pdf' part. So keeping this setting as `false` to make the whole workflow work (without elaborate constructions) is what I'll be doing next. 
1. Also to make things more uniform and more 'ready for change' INCLUDES in the same dir should be managed the same as in other part of the tree. Treating those the same makes it easier in the long run to maintain IMHO
1. Some parts of the workflow (underlying systems) only behave well when absolute paths or relative paths (with parent path backtracking) are used, so we'll just have to use them IMHO
1. the `:parent-path:` attribute declaration in the main document and its use `{parent-path}` everywhere else is what makes this all work
1.  What I didn't test or make use of are nested includes, so that might be something to look into at some point, maybe just overriding the `:parent-path:` attribute is sufficient... There are a lot of rabbit holes to follow ;-)

**EXAMPLE**
VSCODE SETTINGS
```
    "asciidoc.trace": "verbose",
    "asciidoc.use_asciidoctorpdf": true,
    "asciidoc.useWorkspaceRoot": false,
    "asciidoc.preview.attributes": {
      "styles": "",
      "imagesdir": "${workspaceFolder}/resources/media",
      "stylesdir": "${workspaceFolder}/resources/css", // used for preview and in pdfexport when 'use_asciidoctorpdf = false'
      "stylesheet": "my-asciidoc-styles.css" // used for preview and in pdfexport when 'use_asciidoctorpdf = false'
    },
    //"asciidoc.preview.style": "material-green.css",
    "asciidoc.use_asciidoctor_js": true,
    "asciidoc.preview.useEditorStyle": false,
```

MAIN DOCUMENT `Procedures_-_01_Introduction.adoc`
```asciidoc
= Procedures - Introduction
// Set Document Properties
:doctitle: Procedures - Introduction
:author: My Name
// Include Generic Settings
:parent-path: ../../
include::{parent-path}resources/default_settings.adoc[]

toc::[]

include::{parent-path}MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc[leveloffset=1]

include::{parent-path}MyCorp_-_Office_Automation_Notebook/Features/Features_-_50_shared_summary.adoc[leveloffset=1]
...
```

INCLUDE settings in workspaceroot `default_settings.adoc`
```asciidoc
// Path Settings
:imagesdir: {parent-path}resources/media
// Set Theme Configuration (used for PDF export)
:pdf-themesdir: {parent-path}resources/themes
:pdf-theme: my
:pdf-fontsdir: {parent-path}resources/fonts
// Use Local Font (Awesome, Web) fonts
:icons: font
// Store data locally on Export
:data-uri:
:allow-uri-read:
// Set Specific Variables
:showtitle:
...
```

INCLUDE in same dir `Procedures_-_50_shared_glossary.adoc`
```asciidoc
[#_glossary]
= Glossary

We explain the following *terms* which are used throughout this document:
...
```

INCLUDE in other dir `Features_-_50_shared_summary.adoc`
```asciidoc
[#_features]
= Features

We explain the following *features* which are referenced in this document:
...
```

Concluding. To make the experience better it would be nice the `parent-path` attribute would just be calculated as seen from a base-dir (ie. workspace dir) for all aspects in this workflow (both js and asciidoctor-pdf), as you just need to calculate the relative 'depth' of a path and adding `../`for every level. That way I don't have to bother with adding the `parent-path` attribute myself in every document and calculate the relative depth everytime. Just a thought though.

**Comment by ggrossetie** [comment]
> I have reverted back to the release version, because the preview had too many issues (with toc and preview styles not loading period, whatever I did...) -> what follows is based on the current release 2.9.8 version

Could you please create issues (if they don't exist already) on what's not working on 3.x as we will "soon" release a "final" 3.x version?

> Concluding. To make the experience better it would be nice the parent-path attribute would just be calculated as seen from a base-dir (ie. workspace dir) for all aspects in this workflow (both js and asciidoctor-pdf), as you just need to calculate the relative 'depth' of a path and adding ../for every level

The VS Code extension could compute an attribute but then it won't work on the CLI.
The technique you are using is (for now) the recommended approach:

> If you store your AsciiDoc files in nested folders at different levels, relative file paths can quickly become awkward and inflexible. A common pattern to help here is to define the paths in attributes defined in the header, then prefix all include paths with a reference to one of these attributes:

```
:includedir: _includes
:sourcedir: ../src/main/java

include::{includedir}/fragment1.adoc[]

[source,java]
----
include::{sourcedir}/org/asciidoctor/Asciidoctor.java[]
----
```

https://docs.asciidoctor.org/asciidoc/latest/directives/include/#include-resolution

> What I didn't test or make use of are nested includes, so that might be something to look into at some point, maybe just overriding the :parent-path: attribute is sufficient... There are a lot of rabbit holes to follow ;-)

Please keep in mind that, if the include directive is used in a file that has itself been included, the path is resolved relative to the including (i.e., current) file.

**Comment by SjoerdV** [comment]
Great! Eventually it's just a matter of RTFM. Arghh. Nested includes should just work with that strategy, so thanks for that.

Would be nice if these 'best practices' were a bit more easy to find and not somewhere in the middle of a page. Something like [tagging a region](https://docs.asciidoctor.org/asciidoc/latest/directives/include-tagged-regions/) with 'recommended-practice' would be nice? and then collecting all these best-practices and put in one document like here: https://asciidoctor.org/docs/asciidoc-recommended-practices

> The VS Code extension could compute an attribute but then it won't work on the CLI.

Wait... let's just say that I would set `useWorkspaceRoot=true` . Wouldn't it be possible for the extension to update an attribute of my choice (should be an extra setting), if exists, let's say `:parent-path:` in the active document, on save? That way it will also work on the CLI.

**Comment by ggrossetie** [comment]
> Great! Eventually it's just a matter of RTFM. Arghh. Nested includes should just work with that strategy, so thanks for that.

> Would be nice if these 'best practices' were a bit more easy to find and not somewhere in the middle of a page. Something like tagging a region with 'recommended-practice' would be nice? and then collecting all these best-practices and put in one document like here: https://asciidoctor.org/docs/asciidoc-recommended-practices

That's a good idea!
This page has not yet been migrated to the new documentation site but feel free to discuss this change in the project chat: https://chat.asciidoctor.org and submit a pull request.

> Wait... let's just say that I would set useWorkspaceRoot=true...

I still think that setting the `base_dir` option (i.e. not an attribute) on the command line would work too:

```
asciidoctor-pdf --base-dir=/path/to/workspace file.adoc
```

I need to debug a little more on the 3.x version to understand why it's not working.

**Comment by ggrossetie** [comment]
Revisiting this with the current code base (Asciidoctor.js 4.0).

The root cause was a base-directory mismatch: the preview and the `asciidoctor-pdf` export resolved `include::` targets against different directories, so a path written relative to the workspace root worked in the preview but failed in the PDF.

This is now fixed. There is a setting, **`asciidoc.useWorkspaceRootAsBaseDirectory`**, that controls the base directory used to resolve relative includes (and images), and it is applied consistently to *both* the preview and the PDF export:

- **Default (`false`)** — includes resolve relative to the document's own directory, in both the preview and the PDF.
- **Enabled (`true`)** — includes resolve relative to the workspace root, in both the preview and the PDF.

I verified this by parsing/converting a document with the exact options each path uses (the preview via the API, the export via `asciidoctor-pdf -B <base-dir>` reading from stdin), with decoy files at the workspace root: preview and PDF resolve every include — top-level and nested — to the same file in both configurations.

So, for the original use case, enabling `useWorkspaceRootAsBaseDirectory` makes:

```adoc
include::MyCorp_-_Office_Automation_Notebook/Support/Procedures_-_50_shared_glossary.adoc[leveloffset=1]
```

work identically in the preview and the PDF export. The `:parent-path:` workaround is no longer required.

Closing as resolved — please reopen or comment if you still hit a discrepancy. Thanks @SjoerdV for the detailed report!

**Comment by SjoerdV** [comment]
@ggrossetie Thanks for all your great work. Of course I went to try this at once. I did encounter a couple of things that forced me to keep using the `:parent-path:` method for now. 

I use 'multi-homed workspaces'. Ie. in a single `.code-workspace` file I have this structure:

```
  "folders": [
    {
      "path": ".",
      "name": "work"
    },
    {
      "path": "notes/project1",
      "name": "project1"
    },
    {
      "path": "notes/project2",
      "name": "project2"
    }
  ]
```

This gives the option to have a `.vscode` folder inside those 'folders' where for example alternative asciidoc stylesheets can be declared. But... all those folders share the same `/resources` folder on the `${workspaceRoot}` (deprecated variable btw...) (the one with the `.` above) and also most settings are just being inherited by the `.code-workspace` file.

1. The vscode variable convention `${workspaceFolder:<WorkspaceName>}` (in my case `${workspaceFolder:work}`) used in multi-homed workspaces is not being respected both in direct Settings (ie. `"asciidoc.pdf.outputDirectory"`, `"asciidoc.preview.asciidoctorAttributes": { "imagesdir": "${workspaceFolder}/resources/media"}`, etc.) and presumably also in code constructions where the `workspaceFolder` path is used implicitly.

Reference1: https://code.visualstudio.com/docs/editing/workspaces/multi-root-workspaces
Reference2: https://code.visualstudio.com/docs/reference/variables-reference#_variables-scoped-per-workspace-folder

2. If `asciidoc.useWorkspaceRootAsBaseDirectory` is enabled. Which convention do I use for include statements that **do** need to reference a file in the same dir? So `image::image.jpg` helped with `"imagesdir": "${workspaceFolder}/resources/media"`finds the image in the dir in the workspaceRoot, but `include::doc2.adoc` will not be found, because it is in the same folder as the parent doc. I also tried `include::{docdir}/doc2.adoc` but that did not work. But maybe I am missing something obvious here and love to be educated.

**Comment by ggrossetie** [comment]
Thanks @SjoerdV, both points are valid, and both fall outside what the `useWorkspaceRootAsBaseDirectory` fix covered. I've dug into the current code base and split them into follow-up issues so they're easier to track.

**1. `${workspaceFolder:name}` in multi-root workspaces**. You're right, it doesn't work. Two things are going on: VS Code doesn't expand these variables for settings that an extension reads itself, so we substitute them manually. However, today we only handle the
bare `${workspaceFolder}` form (not the named `${workspaceFolder:work}`, nor the deprecated `${workspaceRoot}`), and we even resolve the bare form to different folders in the preview vs. the PDF export. Tracked in #1154.

**2. Same-directory includes when `useWorkspaceRootAsBaseDirectory` is enabled**. This is standard Asciidoctor behaviour: in the *primary* document, relative includes resolve against `base_dir`, so with the base dir set to the workspace root, `include::doc2.adoc`
(a sibling of the main file) is looked up at the root. The intended pattern for that case is to prefix with `{docdir}`:

```adoc
include::{docdir}/doc2.adoc[]
```

The reason `{docdir}` didn't work for you is a real bug on our side. It's set correctly in the preview but **not** passed to `asciidoctor-pdf` in the export, where it falls back to the workspace root. So `{docdir}` currently works in the preview and fails in the PDF —
exactly the kind of divergence this issue was meant to remove. Tracked in #1156.

Separately, in your multi-homed setup the base dir resolves to the sub-folder that contains the document (e.g. `project1`), not the `work` folder that hosts your shared `/resources`. To let you anchor the base dir on a *named* folder, I've opened #1155 to add
a base-directory setting that accepts `${workspaceFolder:work}`.

So for now the `:parent-path:` workaround is still the safe bet; the three issues above should let us remove it properly.
Thanks again for the very precise report.

**Comment by ggrossetie** [comment]
@SjoerdV good news: you can keep your multi-homed layout exactly as it is and retire `:parent-path:`. The right tool here isn't the base-directory setting, it's the `{asciidoctorconfigdir}` attribute together with a `.asciidoctorconfig` file.

#### Recipe

Leave `asciidoc.useWorkspaceRootAsBaseDirectory` **off** (the default), so same-folder includes just work:

```adoc
include::Procedures_-_50_shared_glossary.adoc[]   // sibling, resolved next to the current file
```

Put a `.asciidoctorconfig` at the root of your `work` folder and reference the shared `/resources` through `{asciidoctorconfigdir}` (it resolves to the directory of that config file):

```adoc
# work/.asciidoctorconfig
:assetsdir: {asciidoctorconfigdir}/resources
:imagesdir: {assetsdir}/media
:stylesdir: {assetsdir}/css
:stylesheet: my-asciidoc-styles.css
```

Then in your documents, reference shared includes through that anchor:

```adoc
include::{assetsdir}/default_settings.adoc[]
```

#### Why this fixes it

- **Same-folder includes** resolve naturally, because the base directory stays the document's own folder.
- **Shared assets** resolve through `{asciidoctorconfigdir}`, independent of where the document lives.
- It behaves **identically in the preview and in the extension's PDF export**: the extension prepends the `.asciidoctorconfig` (and defines `asciidoctorconfigdir`) in *both* paths, so the divergence you were hitting is gone.
- In a **multi-root** workspace, a `.asciidoctorconfig` at another folder's root is picked up too (new in the latest release), so the one at `work` applies to documents under `project1` / `project2`.
- `{asciidoctorconfigdir}` is also understood by the IntelliJ AsciiDoc plugin, so it's **portable** across editors, and there's nothing per-document to maintain, unlike `:parent-path:`.

#### Your per-folder `.vscode` / alternative stylesheets

That keeps working, and you can go further with a **per-folder `.asciidoctorconfig`** (e.g. in `notes/project1`) to declare a project-specific stylesheet.

One subtlety, and it's exactly the one you suspected: when several `.asciidoctorconfig` files apply, they are layered general → specific, so **inside your document `{asciidoctorconfigdir}` resolves to the *nearest* one** (here `notes/project1`), not to `work`. So don't reference the shared `/resources` with a bare `{asciidoctorconfigdir}` in the documents when a nearer config exists, it would point at the sub-folder.

The robust pattern is the anchor attribute above: an attribute **value is expanded when it is assigned**, so `:assetsdir: {asciidoctorconfigdir}/resources` defined in `work/.asciidoctorconfig` is frozen to `work/resources` right there and stays correct everywhere, even where a nearer config later re-points `{asciidoctorconfigdir}`. The per-folder config can then override only what's local (e.g. `:stylesheet: {asciidoctorconfigdir}/project.css`, which correctly resolves to *its own* folder).

I verified this layering:

```
# document body
configdir = /ROOT/sub        <- nearest config wins
assets    = /ROOT/resources  <- anchor from the root config, unaffected by the nearer one
style     = /ROOT/sub/...     <- per-folder config, its own dir
```

#### One caveat

`{asciidoctorconfigdir}` is a convention of the editors (this extension and IntelliJ), not of the bare Asciidoctor CLI. If you ever run `asciidoctor-pdf` directly on the command line (outside the extension's export), it won't be defined, pass `-a asciidoctorconfigdir=/path/to/work`. For the extension's export it's set for you.

---

tldr; your structure is fine, no restructuring needed. `{asciidoctorconfigdir}` + a stable anchor attribute in the top-level `.asciidoctorconfig` lets you drop `:parent-path:` while keeping preview and PDF in sync.

**Comment by SjoerdV** [comment]
Awesome! That .asciidoctor file was the missing link indeed. I just switched and removed `:parent-path:` everywhere. All (preview and pdf printing) is working fine and I never use `{asciidoctorconfigdir}` in any document directly so, I'll have no issues.

Thanks for the 'Asciidoctor CLI' tip, I was bound to fall for that one.

As it stands the only 'path-based setting' I still have in vscode settings is `asciidoc.pdf.outputDirectory` which suffers from the multihomed structure. But I see your already have ideas about that.

So I now have this in the toplevel .asciidoctor file
```
// Set Path Settings
:assetsdir: {asciidoctorconfigdir}/resources
:imagesdir: {assetsdir}/media
:stylesdir: {assetsdir}/css
:stylesheet: work-asciidoc-styles.css
// Set Theme Configuration (used for PDF export)
:pdf-themesdir: {assetsdir}/themes
:pdf-theme: my
:pdf-fontsdir: {assetsdir}/fonts,GEM_FONTS_DIR
```

and this in a project level one
```
// Set Path Settings
:stylesheet: project1-asciidoc-styles.css
// Set Theme Configuration (used for PDF export)
:pdf-theme: project1
```

Great stuff!
